### PR TITLE
treewide: use perl.withPackages when possible

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

Since 03eaa48 added perl.withPackages, there is a canonical way to
create a perl interpreter from a list of libraries, for use in script
shebangs or generic build inputs. This method is declarative (what we
are doing is clear), produces short shebangs[1] and needs not to wrap
existing scripts.

Unfortunately there are a few exceptions that I've found:

  1. Scripts that are calling perl with the -T switch. This makes perl
  ignore PERL5LIB, which is what perl.withPackages is using to inform
  the interpreter of the library paths.

  2. Perl packages that depends on libraries in their own path. This
  is not possible because perl.withPackages works at build time. The
  workaround is to add `-I $out/${perl.libPrefix}` to the shebang.

In all other cases I propose to switch to perl.withPackages.

[1]: https://lwn.net/Articles/779997/

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
We might get shorter shebangs from #113800

**Comment by rnhmjoj** [comment]
So, all the changes under `nixos/` and `pkgs/tools/virtualization/nixos-container/` are trivial. The other would require some more accurate testing and possibly a double check from the maintainers.

- [x] nixos/modules/config/users-groups.nix 
- [x] nixos/modules/installer/tools/tools.nix 
- [x] nixos/modules/programs/command-not-found/command-not-found.nix 
- [x] nixos/modules/programs/command-not-found/command-not-found.pl 
- [x] nixos/modules/system/activation/switch-to-configuration.pl 
- [x] nixos/modules/system/activation/top-level.nix 
- [x] nixos/modules/system/boot/loader/grub/grub.nix 
- [x] nixos/modules/system/etc/etc.nix 
- [x] nixos/modules/testing/service-runner.nix 
- [x] pkgs/applications/graphics/feh/default.nix 
- [ ] pkgs/build-support/writers/default.nix 
- [x] pkgs/development/tools/misc/icon-naming-utils/default.nix 
- [ ] pkgs/misc/screensavers/xscreensaver/default.nix 
- [ ] pkgs/os-specific/linux/apparmor/default.nix 
- [ ] pkgs/os-specific/linux/rdma-core/default.nix 
- [ ] pkgs/tools/backup/store-backup/default.nix 
- [ ] pkgs/tools/backup/znapzend/default.nix 
- [ ] pkgs/tools/filesystems/file-rename/default.nix 
- [ ] pkgs/tools/misc/lbdb/default.nix 
- [ ] pkgs/tools/networking/tgt/default.nix 
- [ ] pkgs/tools/package-management/dpkg/default.nix 
- [x] pkgs/tools/virtualization/nixos-container/default.nix 
- [x] pkgs/tools/virtualization/nixos-container/nixos-container.pl

**Review by SuperSandro2000 — COMMENTED** [review]
Took a quick look at the diff and didn't see anything off.

File: pkgs/development/tools/misc/icon-naming-utils/default.nix
Comment (rnhmjoj) [inline pkgs/development/tools/misc/icon-naming-utils/default.nix:13]: I removed the `substituteInPlace` because the shebang is already fixed in the fixup phase.
I assume the hook is picking up the (now right) perl interpreter from the buildInputs.

**Comment by kvtb** [comment]
also https://github.com/NixOS/nixpkgs/commit/8d2f71bda7d4e8e824b52a05b8b69fc7dcd75822

File: pkgs/tools/backup/znapzend/default.nix
Comment (kvtb) [inline pkgs/tools/backup/znapzend/default.nix:26]:
```suggestion
      p.TAPParserSourceHandlerpgTAP
```

`p.IOPipely` is already here via `MojoIOLoopForkCall-0-20`

`perl.withPackages` magically brings all the deps which one had to explicitly put into `$PERL5LIB`.
There could be more places like that (probably all of the `$PERL5LIB`->`perl.withPackages` replacements)

Comment (kvtb) [inline pkgs/tools/backup/znapzend/default.nix:26]:
```suggestion
    [ MojoIOLoopForkCall-0-20 p.TAPParserSourceHandlerpgTAP
```


`p.IOPipely` and `Mojolicious-8-35` are already here via `MojoIOLoopForkCall-0-20`

`perl.withPackages` magically brings all the deps which one had to explicitly put into `$PERL5LIB`.
There could be more places like that (probably all of the `$PERL5LIB`->`perl.withPackages` replacements)

File: nixos/modules/programs/command-not-found/command-not-found.nix
Comment (kvtb) [inline nixos/modules/programs/command-not-found/command-not-found.nix:18]:
```suggestion
    perl = pkgs.perl.withPackages (p: [ p.DBDSQLite p.StringShellQuote ]);
```

File: pkgs/tools/backup/znapzend/default.nix
Comment (rnhmjoj) [reply pkgs/tools/backup/znapzend/default.nix]: Oh right, thank you for the remainder. Fixed it.

File: nixos/modules/programs/command-not-found/command-not-found.nix
Comment (rnhmjoj) [reply nixos/modules/programs/command-not-found/command-not-found.nix]: Done.

**Comment by rnhmjoj** [comment]
I added volth's (yours?) commits as well. Thank you.
I tested nix-serve and loading `nix.perl-bindings` in a nix-shell and both works.

**Comment by rnhmjoj** [comment]
Pinging maintainers:

- xscreensaver: @7c6f434c
- apparmor: @phreedom @thoughtpolice @joachifm
- rdma-core: @markuskowa
- store-backup: @marcweber
- znapzend: @otwieracz
- file-rename: @peterhoeg
- ldbd: @kaiha @bfortz
- tgt: @johnazoidberg
- dpkg: @siriobalmelli

I'm doing some non-trivial changes (related to perl) to you packages. It would help a lot if you could check this doesn't break anything. Thank you.

File: pkgs/os-specific/linux/rdma-core/default.nix
Comment (markuskowa) [inline pkgs/os-specific/linux/rdma-core/default.nix:42]: Looking at the output, it seems to work. However, shouldn't the modified Perl environment be in `buildInputs`, such that the `fixupPhase` can automatically fix the shebang?

Comment (rnhmjoj) [reply pkgs/os-specific/linux/rdma-core/default.nix:42]: Ideally yes, but I can't add `$out/${perl.libPrefix}` using perl.withPackages. This is one of the exceptional cases I mentioned.

File: pkgs/misc/screensavers/xscreensaver/default.nix
Comment (7c6f434c) [inline pkgs/misc/screensavers/xscreensaver/default.nix:47]: Make sense / localise enough (to be honest, I am not sure I have ever tried to enable this functionality)

**Review by bfortz — APPROVED** [review]
pkgs/tools/misc/lbdb/default.nix works for me.

File: pkgs/tools/backup/znapzend/default.nix
Comment (kvtb) [inline pkgs/tools/backup/znapzend/default.nix:64]: I have not checked, but does `substituteInPlace` do anything here?
Since we have `perl'` instead of `perl` in `buildInputs = []`, how `${perl}/bin/perl` could appear inside the files?
`patchShebang .` is to put `${perl'}/bin/perl` there, so `substituteInPlace` seems unnecessary

Comment (rnhmjoj) [reply pkgs/tools/backup/znapzend/default.nix]: Right, it's actually useless: I have just checked.

File: pkgs/misc/screensavers/xscreensaver/default.nix
Comment (kvtb) [inline pkgs/misc/screensavers/xscreensaver/default.nix:47]:
```suggestion
         --replace '${perl}' '${perl.withPackages (p: [ p.LWPProtocolHttps ])}'
```

the list is merely deps of LWPProtocolHttps

Comment (rnhmjoj) [reply pkgs/misc/screensavers/xscreensaver/default.nix]: Thank you, fixed.

**Comment by siriobalmelli** [comment]
Built and tested `dpkg` on Darwin, LGTM

**Comment by peterhoeg** [comment]
I haven't tried it out, but the `file-rename` change looks perfectly reasonable. Thanks @rnhmjoj

**Comment by rnhmjoj** [comment]
Fixed a couple conflicts.

**Comment by aanderse** [comment]
@stigtsp and @volth (does anyone know their handle these days?) are we good to merge this?

**Comment by rnhmjoj** [comment]
Most of the packeges have been reviewed. I'm running now some basic NixOS tests but the changes should be ok.
Two questions before merging:

1. what's the EditorConfig check and why is it failing?
2. considering the numebr of rebuilds, staging or master?

**Comment by stigtsp** [comment]
> are we good to merge this?

Diff LGTM.

**Comment by rnhmjoj** [comment]
I run some NixOS installer and bootloader tests: all good.

**Review by stigtsp — APPROVED** [review]

**Comment by rnhmjoj** [comment]
Ok, fixed the EditorConfig check: there was a trailing whitespace in one of volth's commit.

**Comment by rnhmjoj** [comment]
I changed the base to staging because it's not so clear whether this PR will cause many slow rebuilds.

**Pull request merged**
